### ISSUE TITLE
losses are nan

### ISSUE THREAD

**Issue by PoYuHan** [ultralytics/yolov5#4084]
Hi, recently I trained my own data with new train.py, all losses I got were nan and the confusion matrix shows every images were predicted as FN, I also tried on coco128 datasets and thing still goes wrong, but another PC training by CPU seems good, and the older version train.py training in GPU works good to.

OS: Windows 10
torch:1.9.0-cuda11.1

And sorry for my broken English...

**Comment by glenn-jocher** [comment]
@PoYuHan sometimes Windows and/or Anaconda environments suffer from CUDA problems.

It appears you may have environment problems. Please ensure you meet all dependency requirements if you are attempting to run YOLOv5 locally.  If in doubt, create a new virtual Python 3.8 environment, clone the latest repo (**code changes daily**), and `pip install -r requirements.txt` again. We also highly recommend using one of our verified environments below. 


## Requirements

[**Python>=3.6.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.7**](https://pytorch.org/get-started/locally/). To get started:
```bash
$ git clone https://github.com/ultralytics/yolov5
$ cd yolov5
$ pip install -r requirements.txt
```

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), validation ([val.py](https://github.com/ultralytics/yolov5/blob/master/val.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/export.py)) on MacOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by PoYuHan** [comment]
I used anaconda to create a python3.8 env and follow all steps in install section of Quick Start Examples just 3 hours ago, but still not work.

**Comment by glenn-jocher** [comment]
@PoYuHan we recommend pip envs and pip installs due to problems like this with Anaconda.

**Comment by PoYuHan** [comment]
I've tried to use pipenv and install all packages with requirements.txt, and losses calculated correctly but it was using CPU, so I manually reinstall pytorch, torchvision, and torchaudio the cuda 11.1 version follows by the Pytorch official website, after that it used GPU but all losses turned to nan again.

**Comment by glenn-jocher** [comment]
@PoYuHan then your data is causing training instabilities and you should check it for errors.

If you have a reproducible issue on a common dataset like COCO128 please raise a new issue. We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

* ✅ **Minimal** – Use as little code as possible that still produces the same problem
* ✅ **Complete** – Provide **all** parts someone else needs to reproduce your problem in the question itself
* ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

In addition to the above requirements, for [Ultralytics](https://ultralytics.com/) to provide assistance your code should be:

* ✅ **Current** – Verify that your code is up-to-date with current GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been resolved by previous commits.
* ✅ **Unmodified** – Your problem must be reproducible without any modifications to the codebase in this repository. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all of the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) and providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by PoYuHan** [comment]
@glenn-jocher I've tried uninstall anaconda and reinstall python then use pipenv to train, and everything works great!! Thanks!!

**Comment by glenn-jocher** [comment]
@PoYuHan oh good! So the problem at the end was anaconda then?

Maybe we should have a warning to avoid anaconda installs.

**Comment by PoYuHan** [comment]
@glenn-jocher I think I was wrong... I just opened my pipenv then do the training and error still appears, I tried rebuild the virtrual env and it was not helping. I've also tried on another pc but the same error happened.

**Comment by PoYuHan** [comment]
@glenn-jocher I tried installed cuda10.2 and pytorch=1.9.0+cuda102 then the problem solved!! I also tried anaconda and pipenv, both of them worked perfectly on cuda10.2 version. Thank you so much for your help!!!!!!!!

**Comment by mengmeng0406** [comment]
> 

@PoYuHan 
I encountered the same problem. How did you solve it? Just change CUDA version to 10.2? The CUDA version of my virtual environment is 10.2, but the CUDA version of the system is 11 x.

**Comment by shubhambagwari** [comment]
I forked the latest version of Yv5, Still, I am facing this error.
Earlier I trained this model with 2000 images but faced the same issue.

Roboflow has been used to label the custom dataset.
I have read the earlier discussions but not working.

![207236797-e2cf2e4d-67ec-4a52-8fdb-ff048cf0caa3](https://user-images.githubusercontent.com/25363312/207513725-6c5e88ca-b841-4645-aeff-7eff94c4608f.png)

**Comment by glenn-jocher** [comment]
@shubhambagwari train 300 epochs.

**Comment by rtoddsullivan** [comment]
I'm running on wsl2 using python env configured with requirements.txt. I find that when I define a large batch size, train_loss and test_loss are `nan`. 
For me a batch size of 16 works but 128 shows `nan`.

**Comment by glenn-jocher** [comment]
@rtoddsullivan This issue might be related to numerical instabilities when using large batch sizes. Using mixed precision training (torch.backends.cudnn.benchmark = True) can sometimes alleviate this problem without sacrificing performance.

You can also try reducing the learning rate and adjusting the network architecture, if possible, to stabilize training with larger batch sizes.

Thank you for your contribution!
